include/pybind11/pytypes.h

```diff
@@ -377,20 +380,49 @@ class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
     /// Constructs a new exception from the current Python error indicator, if any.  The current
     /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    error_already_set() : std::runtime_error("") {
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        if (m_type) {
+            PyErr_NormalizeException(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        }
     }
 
     error_already_set(const error_already_set &) = default;
     error_already_set(error_already_set &&) = default;
 
     inline ~error_already_set() override;
 
+    const char *what() const noexcept override {
+        if (m_lazy_what.empty()) {
+            try {
+                m_lazy_what = detail::error_string(m_type.ptr(), m_value.ptr(), m_trace.ptr());
+            } catch (const std::exception &e) {
+                m_lazy_what
+                    = std::string(
+                          "Unknown internal error occurred while constructing error_string: ")
+                      + e.what();
```

Comment:
This looks like it could result in infinite recursion: we're using things like `str` and `cast` in `error_string`, which may throw `error_already_set`.

The more I look at this, the more uneasy I feel, unfortunately.

That could be alleviated by radically simplifying, as you suggest also IIUC, starting with:
```
try {
    m_lazy_what = detail::error_string(m_type.ptr(), m_value.ptr(), m_trace.ptr());
} catch (...) {
    std::cerr << "FATAL failure building pybind11::error_already_set error_string (" << __FILE__ << ":" << __LINE__ << ")" << std::endl;
    std::terminate();
}
```

But I'd also want to print out information about `m_type`, `m_value` if possible, although I'd be OK doing that in a separate PR.

Reply:
Spooked by the fear that original errors are masked by errors in the error handling, I worked out a robust alternative. I just pushed a change that generates the output below. The implementation is pretty straightforward, meant to maximize reliability, which make it a bit cumbersome.

The output includes:
1. The Python exception type & value (if available).
2. The C++ exception that was thrown when building error_string.

I believe 2. may not appear on all platforms, but at least the Python error will always appear. 

```
Running tests in directory "/usr/local/google/home/rwgk/forked/pybind11/tests":
=========================================================== test session starts ============================================================
platform linux -- Python 3.9.10, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /usr/local/google/home/rwgk/forked/pybind11/tests, configfile: pytest.ini
collected 537 items

test_async.py ..                                                                                                                     [  0%]
test_buffers.py .........                                                                                                            [  2%]
test_builtin_casters.py .FATAL failure building pybind11::error_already_set error_string: UnicodeDecodeError: "'utf-8' codec can't decode byte 0xd0 in position 3: invalid continuation byte"
terminate called after throwing an instance of 'std::runtime_error'
  what():  Something went wrong.
Fatal Python error: Aborted

Current thread 0x00007ff12f6ef740 (most recent call first):
  File "/usr/local/google/home/rwgk/forked/pybind11/tests/test_builtin_casters.py", line 24 in test_unicode_conversion
  File "/usr/local/lib/python3.9/dist-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/usr/local/lib/python3.9/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.9/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.9/dist-packages/_pytest/python.py", line 1641 in runtest
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/usr/local/lib/python3.9/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.9/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 255 in <lambda>
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 311 in from_call
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 215 in call_and_report
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/usr/local/lib/python3.9/dist-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/usr/local/lib/python3.9/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.9/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.9/dist-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/usr/local/lib/python3.9/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.9/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.9/dist-packages/_pytest/main.py", line 323 in _main
  File "/usr/local/lib/python3.9/dist-packages/_pytest/main.py", line 269 in wrap_session
  File "/usr/local/lib/python3.9/dist-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/usr/local/lib/python3.9/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/local/lib/python3.9/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.9/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.9/dist-packages/_pytest/config/__init__.py", line 162 in main
  File "/usr/local/lib/python3.9/dist-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/usr/local/lib/python3.9/dist-packages/pytest/__main__.py", line 5 in <module>
  File "/usr/lib/python3.9/runpy.py", line 87 in _run_code
  File "/usr/lib/python3.9/runpy.py", line 197 in _run_module_as_main
```